Add weapon strafing support and implement it on RA Yak

reaperrr opened the pull request:
This has been overdue in my opinion.

Modifying the targeted offset per burst is by far the easiest approach compared to modifying Attack traits or projectiles.

Closes #2642.
Fixes #2730.

reaperrr committed:
Add strafing support to weapons

reaperrr committed:
Use strafing for RA Yak

This is done by modifying the target offset by a value matching the Yak's
speed * BurstDelay, to sync the offsets with the actor speed.

reaperrr commented:
Pinging @SoScared and @OpenRA/engine-hackers for feedback on whether we might want this for the playtest (or not at all, at least wrt using it in RA).

reaperrr commented:
Moving it to playtest for now, as it would be best to include it together with the hit-shape changes so balance can be revamped in one go.

pchote commented:
Needs a rebase.

pchote reviewed (commented):
Two things I noticed while testing:
* The yak now (usually) empties its entire ammo pool in the first past, instead of spreading it over two passes.  We should keep it at two passes before reloading.
* Sometimes it only fires one blip in the first pass.  Probably related to #13220 and so out of scope of this PR, but still annoying.

Otherwise, the code implementation looks good.

It would be good if you could remove the hardcoded strafing logic from `AttackBomber` and make it use this instead.  It should be possible to remove the custom `Guns` and `Bombs` properties and use `Armaments` from `AttackBase`.

reaperrr committed:
Reset burst counter if ReloadDelay has passed since last shot

This fixes the following issues:
- units like mammoth tanks sometimes at first only fired 1 shot on new encounter because they only depleted 1 burst before previous target was killed
- weapons that use strafing logic would not reset the offset multiplier after passing the target once, leading to wrong offsets on following attacks

reaperrr commented:
Rebased and updated.

I think I fixed `Sometimes it only fires one blip in the first pass. Probably related to #13220 and so out of scope of this PR, but still annoying.`, though I didn't fully took the approach suggested in #13220 as completely replacing the FireDelay/ReloadDelay dance would require covering some edge cases and I didn't want to risk more regressions than necessary.

> It would be good if you could remove the hardcoded strafing logic from AttackBomber and make it use this instead. It should be possible to remove the custom Guns and Bombs properties and use Armaments from AttackBase.

I'll see how far I get on this.

reaperrr committed:
AttackBomber style fixes

reaperrr committed:
Enable RA ParaBomb shadow

reaperrr committed:
Unhardcode AttackBomber bomb and gun logic

reaperrr commented:
Updated, ripped out the hardcoded `AttackBomber` strafing logic.

pchote commented on OpenRA.Mods.Common/Traits/Armament.cs:
```diff
@@ -221,10 +225,32 @@ public virtual Barrel CheckFire(Actor self, IFacing facing, Target target)
+			{
+				// We want this to match Armament.LocalOffset, so we need to convert it to forward, right, up
+				initialOffset = new WVec(initialOffset.Y, -initialOffset.X, initialOffset.Z);
+				passiveTarget += initialOffset.Rotate(WRot.FromFacing(legacyFacing));
```
`legacyFacing` is the facing of the barrel, which is not necessarily the same as the facing of the actor.  This makes sense I guess, but might not be what you expect.

reaperrr replied:
If you're fine with it, I'd actually like to leave it like this. If I or any other modder ever needs a different behavior, we can still fix it (or make it configurable).

pchote commented on mods/cnc/weapons/other.yaml:
```diff
@@ -65,7 +65,7 @@ Chemspray:
 	Inherits: ^FlameWeapon
 	ReloadDelay: 4
-	Range: 2c0
+	Range: 5c0
```
What is this for?  Doesn't seem related to this PR.

reaperrr replied:
I thought this was necessary to make the aircraft start dropping bombs early enough, but I admit I don't remember testing it.

pchote commented on mods/ra/weapons/smallcaliber.yaml:
```diff
@@ -201,16 +201,24 @@ ChainGun:
-	ReloadDelay: 3
-	Range: 5c0
-	MinRange: 3c0
+	Burst: 5
```
This won't evenly divide into 18, but nothing will so we can leave this for a followup balance patch to hopefully iron out.

reaperrr replied:
In my opinion the most sensible fix is (probably) to increase ammo to 20, I didn't want to make that call on my own.

pchote reviewed (commented):
Looks good.  Just one blocking question / probably change request, and a couple of info points:

reaperrr committed:
TD airstrike balance tweak

The increased building hitshape sizes allowed TD airstrikes to kill even Cyards and temples on bleed, which is overpowered compared to before the hitshapes were merged.
Cutting bomb damage vs. Wood to 65% lets Cyards, temples and adv. comm. centres survive airstrikes with yellow health.

reaperrr committed:
RA Parabombs balance fix

They're too over-powered on bleed due to the hitshap changes.

reaperrr commented:
Updated, TD airstrike now behaves pretty much like on bleed again.

Additionally, I added balance commits for TD airstrike and RA parabomb strike, because on bleed they are so obviously over-powered vs. larger building shapes now, that not making changes before the playtest would probably just result in testers focusing too much on those airstrikes.
These changes are obviously not meant to be a final solution, they just make sure that airstrikes don't 1HKO cyards.

pchote commented on OpenRA.Mods.Common/Traits/Armament.cs:
```diff
@@ -169,6 +171,8 @@ protected virtual void Tick(Actor self)
 			if (IsTraitDisabled)
 				return;
 
+			++ticksSinceLastShot;
```
Style nit: Try to avoid incrementing an integer without bound because it could in principle overflow and become negative.  Thats not likely in practice to happen here, but it is still bad style.  You can put this behind an `if (ticksSinceLastShot < Weapon.ReloadDelay)`.

pchote commented on OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs:
```diff
@@ -749,6 +749,22 @@ internal static void UpgradeActorRules(ModData modData, int engineVersion, ref L
+						{
+							Console.WriteLine("Hardcoded Guns and Bombs logic has been removed from AttackBomber.");
+							Console.WriteLine("Bombs should work like usual, for gun strafing use the new Weapon TargetOffset modifiers.");
+							Console.WriteLine("Look at the C&C1 mod's A10 for an example.");
```
Style nit: C&C1 &rarr; TD.

pchote reviewed (approved):
Looks good :+1:. Just a couple of minor style nits:

reaperrr commented:
Fixed.

FrameLimiter commented:
Just gave this a quick test and noticed the damage vs infantry is far too low.  It takes two full strafes (full ammo payload) to kill just 1 Flame thrower (40hp).  Also killing artillery with yaks is far less effective now; before you could kill four lined up artillery in one pass, but now you can only kill 1 of them.

I think both the strafe damage & the ability to queue up multiple targets could use some tweaking.

reaperrr committed:
Rebalanced Yak

- increased ammo per gun from 9 to 10, for two full double-5-burst salvoes, adjusted ReloadDelay to keep total reload time about the same
- increased Spread by 33% to make up for lower effective pin-point accuracy
- increased damage vs. infantry by 25%, vs. light armor by 16.67%, vs. heavy armor by ~10%

reaperrr commented:
@FrameLimiter I added some preliminary balance changes, please re-test. Note that they don't need to be perfect for the playtest, just 'good enough', though.

abcdefg30 commented:
Needs a rebase again.

abcdefg30 commented:
I'm really not sure if we want this enabled on the  yak. You can't easily kill two arties in one run anymore.
That aside 👍 / ✅ .

reaperrr commented:
Rebased.

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/0a4ac6bb1981753dbbf3e017ae6a51ded1a29061...aad16a0588898957ab59b987c5b2946e7b59e587)

SoScared commented:
Late to the party but my thoughts are as follows:

I've tried to keep an open mind but I'm not sure the Yak change is worth pandering towards the original game in this case. The repurposing is blind and I don't see a relevant problem with #2642 #2730. The RA mod meta has developed heavily around the current Yak and I'm not sure if it's wise to bring this aboard the official playtest if we want players to dive into the HitShape/UnitStance seriously.
